### ISSUE TITLE
analysis: fetching incorrect bitness for segments on elf64

### ISSUE THREAD

**Issue by williballenthin** [williballenthin/python-idb#28]
for sample 4F428F95740EC935125766DE5BAA8A0D, all segments are 64-bit, with bitness == 2:

```
Python>[idc.GetSegmentAttr( s, idc.SEGATTR_BITNESS) for s in idautils.Segments()]
[2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L]
```

but fetching segment bitness is returning 3 (as reported by @emdel).

**Comment by williballenthin** [comment]
will add test case as soon as possible

**Comment by williballenthin** [comment]
A9828BE4ADC636E7470EC985FCF5D5F9 is `/bin/cat`.

```
Python>[idc.GetSegmentAttr( s, idc.SEGATTR_BITNESS) for s in idautils.Segments()]
[2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L, 2L]
Python>GetInputFileMD5()
A9828BE4ADC636E7470EC985FCF5D5F9
```

**Comment by williballenthin** [comment]
for `/bin/cat` here's the IDA 7.0 output of `dump_section_list.py`:

```
LOAD
 - start address: 0x400000
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.init
 - start address: 0x401598
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x4
 - type: 0x2
 - color: 0xffffffff
LOAD
 - start address: 0x4015b2
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.plt
 - start address: 0x4015c0
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x3
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x5
 - type: 0x2
 - color: 0xffffffff
.text
 - start address: 0x401a40
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x3
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x6
 - type: 0x2
 - color: 0xffffffff
LOAD
 - start address: 0x40898a
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.fini
 - start address: 0x40898c
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x7
 - type: 0x2
 - color: 0xffffffff
LOAD
 - start address: 0x408995
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.rodata
 - start address: 0x4089a0
 - sclass: 0x9
 - orgbase: 0x0
 - align: 0x8
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x8
 - type: 0x3
 - color: 0xffffffff
LOAD
 - start address: 0x409af3
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.eh_frame_hdr
 - start address: 0x409af4
 - sclass: 0x9
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x9
 - type: 0x3
 - color: 0xffffffff
.eh_frame
 - start address: 0x409e00
 - sclass: 0x9
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xa
 - type: 0x3
 - color: 0xffffffff
.init_array
 - start address: 0x60ae10
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xb
 - type: 0x3
 - color: 0xffffffff
.fini_array
 - start address: 0x60ae18
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xc
 - type: 0x3
 - color: 0xffffffff
.jcr
 - start address: 0x60ae20
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xd
 - type: 0x3
 - color: 0xffffffff
LOAD
 - start address: 0x60ae28
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x2
 - type: 0x3
 - color: 0xffffffff
.got
 - start address: 0x60aff8
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xe
 - type: 0x3
 - color: 0xffffffff
.got.plt
 - start address: 0x60b000
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0xf
 - type: 0x3
 - color: 0xffffffff
LOAD
 - start address: 0x60b250
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x2
 - type: 0x3
 - color: 0xffffffff
.data
 - start address: 0x60b260
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0x8
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x10
 - type: 0x3
 - color: 0xffffffff
LOAD
 - start address: 0x60b314
 - sclass: 0x3
 - orgbase: 0x0
 - align: 0x1
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x2
 - type: 0x3
 - color: 0xffffffff
.bss
 - start address: 0x60b320
 - sclass: 0x13
 - orgbase: 0x0
 - align: 0x8
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x11
 - type: 0x9
 - color: 0xffffffff
extern
 - start address: 0x60bce8
 - sclass: 0x0
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x0
 - bitness: 0x2
 - flag: 0x10
 - sel: 0x12
 - type: 0x1
 - color: 0xffffffff
```

and here's the output when run under python-idb:

```
.init
 - start address: 0x80496ac
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x1
 - type: 0x2
 - color: 0xffffffff
.plt
 - start address: 0x80496d0
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x3
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x2
 - type: 0x2
 - color: 0xffffffff
.plt.got
 - start address: 0x8049de0
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0xa
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x3
 - type: 0x2
 - color: 0xffffffff
.text
 - start address: 0x8049df0
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x3
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x4
 - type: 0x2
 - color: 0xffffffff
.fini
 - start address: 0x805b634
 - sclass: 0x2
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x5
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x5
 - type: 0x2
 - color: 0xffffffff
.rodata
 - start address: 0x805b660
 - sclass: 0x8
 - orgbase: 0x0
 - align: 0x8
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x6
 - type: 0x3
 - color: 0xffffffff
.eh_frame_hdr
 - start address: 0x8060c14
 - sclass: 0x8
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x7
 - type: 0x3
 - color: 0xffffffff
.eh_frame
 - start address: 0x8061430
 - sclass: 0x8
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x4
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x8
 - type: 0x3
 - color: 0xffffffff
.init_array
 - start address: 0x8067f00
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x9
 - type: 0x3
 - color: 0xffffffff
.fini_array
 - start address: 0x8067f04
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xa
 - type: 0x3
 - color: 0xffffffff
.jcr
 - start address: 0x8067f08
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xb
 - type: 0x3
 - color: 0xffffffff
.got
 - start address: 0x8067ffc
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xc
 - type: 0x3
 - color: 0xffffffff
.got.plt
 - start address: 0x8068000
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x5
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xd
 - type: 0x3
 - color: 0xffffffff
.data
 - start address: 0x80681e0
 - sclass: 0xc
 - orgbase: 0x0
 - align: 0x8
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xe
 - type: 0x3
 - color: 0xffffffff
.bss
 - start address: 0x8068380
 - sclass: 0x13
 - orgbase: 0x0
 - align: 0x9
 - comb: 0x2
 - perm: 0x6
 - bitness: 0x1
 - flag: 0x10
 - sel: 0xf
 - type: 0x9
 - color: 0xffffffff
extern
 - start address: 0x8068fb8
 - sclass: 0x0
 - orgbase: 0x0
 - align: 0x3
 - comb: 0x2
 - perm: 0x0
 - bitness: 0x1
 - flag: 0x10
 - sel: 0x10
 - type: 0x1
 - color: 0xffffffff
```

**Comment by williballenthin** [comment]
by making the segment fields `name_index`, `sclass`, and `orgbase` pointer-sized (and therefore dependent on `wordsize`), the two segment lists match one another.

**Comment by williballenthin** [comment]
no regression in other unit tests. looks like this may be an appropriate fix.

**Comment by williballenthin** [comment]
this issue is fixed in bc6de58ee12bd165248a41bd2ffc6f3e218dd6f4

this was an issue in parsing metadata about segments. the size of some fields was dependent on the database wordsize. turns out we only had segment metadata tests for 32-bit databases, so the bug escaped us. 

thanks to @emdel for reporting this issue and taking an initial stab at fixing it!
